x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/execution/search/Querier.java

```diff
@@ -122,23 +131,60 @@ public void query(List<Attribute> output, QueryContainer query, String index, Ac
         List<Tuple<Integer, Comparator>> sortingColumns = query.sortingColumns();
         listener = sortingColumns.isEmpty() ? listener : new LocalAggregationSorterListener(listener, sortingColumns, query.limit());
 
-        ActionListener<SearchResponse> l = null;
-        if (query.isAggsOnly()) {
+        if (cfg.task() != null && cfg.task().isCancelled()) {
+            listener.onFailure(new TaskCancelledException("cancelled"));
+        } else if (query.isAggsOnly()) {
+            ActionListener<SearchResponse> l;
             if (query.aggs().useImplicitGroupBy()) {
                 l = new ImplicitGroupActionListener(listener, client, cfg, output, query, search);
             } else {
                 l = new CompositeActionListener(listener, client, cfg, output, query, search);
             }
+            client.search(search, l);
         } else {
-            search.scroll(cfg.pageTimeout());
-            l = new ScrollActionListener(listener, client, cfg, output, query);
+            searchWithPointInTime(search, new SearchHitActionListener(listener, client, cfg, output, query, sourceBuilder));
         }
+    }
 
-        if (cfg.task() != null && cfg.task().isCancelled()) {
-            listener.onFailure(new TaskCancelledException("cancelled"));
-            return;
+    private void searchWithPointInTime(SearchRequest search, ActionListener<SearchResponse> listener) {
```

Comment:
👍 will give it a try. I also like the idea of having the query and PIT management logic in one place.

Reply:
After having a closer look it seems like that would require some larger refactorings in EQL with only limited payoff. The main difference is that EQL only maintains state within a request and SQL has to carry some state across scroll requests. Since `QueryClient` encapsulates some of the scroll state it would have to be serialized in the cursor but that's not feasible because it also includes the ES client.

I would prefer to keep larger changes to EQL out of this PR and consider the refactoring for a later steps (if necessary).